contracts/src/WarpProtocolRegistry.sol

```diff
@@ -0,0 +1,181 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import "@subnet-evm-contracts/interfaces/IWarpMessenger.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+
+/**
+ * @dev Implementation of an abstract `WarpProtocolRegistry` contract.
+ *
+ * This implementation is a contract that can be used as a base contract for protocols that are
+ * build on top of warp. It allows the protocol to be upgraded through a warp out-of-band message.
+ */
+abstract contract WarpProtocolRegistry {
+    // Address that the out-of-band warp message sets as the "source" address.
+    // The address is obviously not owned by any EOA or smart contract account, so it
+    // can not possibly be the source address of any other warp message emitted by the VM.
+    address public constant VALIDATORS_SOURCE_ADDRESS = address(0);
+
+    WarpMessenger public constant WARP_MESSENGER =
+        WarpMessenger(0x0200000000000000000000000000000000000005);
+
+    bytes32 internal immutable _chainID;
+
+    // The latest protocol version. 0 means no protocol version has been added, and isn't a valid version.
+    uint256 internal _latestVersion;
+
+    mapping(uint256 => address) internal _versionToAddress;
+
+    /**
+     * @dev Emitted when a new protocol version is added to the registry.
+     */
+    event AddProtocolVersion(
+        uint256 indexed version,
+        address indexed protocolAddress
+    );
+
+    // Errors
+    error InvalidWarpMessage();
+    error InvalidSourceChainID();
+    error InvalidOriginSenderAddress();
+    error InvalidDestinationChainID();
+    error InvalidDestinationAddress();
+    error InvalidProtocolAddress();
+    error InvalidProtocolVersion();
+    error InvalidRegistryInitialization();
+
+    /**
+     * @dev Initializes the contract by setting a `chainID` and `latestVersion`.
+     */
+    constructor(
+        uint256[] memory initialVersions,
+        address[] memory initialProtocolAddresses
+    ) {
+        _latestVersion = 0;
+        _chainID = WARP_MESSENGER.getBlockchainID();
+        if (initialProtocolAddresses.length != initialVersions.length) {
+            revert InvalidRegistryInitialization();
+        }
+
+        for (uint256 i = 0; i < initialVersions.length; i++) {
+            _addToRegistry(initialVersions[i], initialProtocolAddresses[i]);
+        }
+    }
+
+    /**
+     * @dev Gets and verifies a warp out-of-band message, and adds the new protocol version
+     * addres to the registry.
+     *
+     * Emits a {AddProtocolVersion} event when successful.
+     * Requirements:
+     *
+     * - a valid warp out-of-band message must be provided.
+     * - the version must be the increment of the latest version.
+     * - the protocol address must be a contract address.
+     */
+    function addProtocolVersion(uint32 messageIndex) external {
```

Comment:
How do we differentiate an out-of-band Warp message from a generic Warp message? I may be missing something, but it seems like this could be called by anyone via Warp.

Reply:
discussed in standup, but it's based on our own requirements for interacting with registry. In this case, sender is `address(0)` which can't be the case for regular warp message, also checking source/destination chain IDs same.